DatadogInternal/Sources/NetworkInstrumentation/URLSession/URLSessionTaskInterception.swift

```diff
@@ -97,6 +97,30 @@ public struct ResourceCompletion {
     }
 }
 
+/// An immutable version of `URLRequest`.
+///
+/// Introduced in response to concerns raised in https://github.com/DataDog/dd-sdk-ios/issues/1638 
+/// it makes a copy of request attributes, safeguarding against potential thread safety issues arising from concurrent 
+/// mutations (see more context in https://github.com/DataDog/dd-sdk-ios/pull/1767 ).
+public struct ImmutableRequest {
+    /// The URL of the request.
+    public let url: URL?
+    /// The HTTP method of the request.
+    public let httpMethod: String?
+    /// The HTTP header fields of the request.
+    public let allHTTPHeaderFields: [String: String]?
+    /// A reference to the original `URLRequest` object provided during initialization. Direct use is discouraged
+    /// due to thread safety concerns. Instead, necessary attributes should be accessed through `ImmutableRequest` fields.
+    public let unsafeOriginal: URLRequest
```

Comment:
we shouldn't keep a copy here else we would end up having the same issue again unknowingly.

Reply:
Agree 👍, however we provide `URLRequest` in RUM Attributes Provider API, which requires passing the original request [here](https://github.com/DataDog/dd-sdk-ios/blob/5334eab9cb5f9de2f47f62d9a951b3b05230938e/DatadogRUM/Sources/Instrumentation/Resources/URLSessionRUMResourcesHandler.swift#L102) in `DatadogRUM`. I share the same concern - we may overlook it and add a new usage somewhere else. To mitigate this risk, I named this property as `unsafeOriginal` so it should raise attention. As an extra layer, we could add linter rule to prevent more usage, but even this can be muted.

Reply:
I think we have made interesting choices with `rumAttributesProvider` API and now we are hand locked. Thanks for background.